src/liballoc/rc.rs

```diff
@@ -570,15 +570,24 @@ impl<T: ?Sized> Rc<T> {
         ptr
     }
 
-    /// Constructs an `Rc` from a raw pointer.
+    /// Constructs an `Rc<T>` from a raw pointer.
     ///
-    /// The raw pointer must have been previously returned by a call to a
-    /// [`Rc::into_raw`][into_raw].
+    /// The raw pointer must have been previously returned by a call to
+    /// [`Rc<U>::into_raw`][into_raw] where `U` must have the same size
+    /// and alignment as `T`. This is trivially true if `U` is `T`.
```

Comment:
Is same size and align strong enough? What stops rustc from ordering the `strong` and `weak` fields of RcBox and ArcInner in a different order for different T?

Reply:
Technically, nothing, but we know today that rustc will not do so I believe. I continue to believe that it would be a good idea to add `#[repr(C)]` or something like it to RcBox and ArcInner, since there should be no reason to not guarantee (internally at least) that the field layout is consistent.

Notably, the data field *must* be last, due to unsizing, and reodering two usize fields is not currently ever going to happen anyway, I believe.

Reply:
> reordering two usize fields is not currently ever going to happen anyway, I believe

Currently it doesn't, but https://github.com/rust-lang/unsafe-code-guidelines/issues/36 is about whether to make this a guarantee, and I don't think it's decided yet. For example PGO-based field layout would absolutely reorder fields with the same size and align to keep frequently read data together and frequently written data in a different cache line.

I don't necessarily intend on blocking this PR on this, but I would feel more comfortable with #\[repr(C)\] on those types.

Reply:
I would be in favor of adding #[repr(C)] as well. Doing so in this PR makes sense to me.

Reply:
Amended `RcBox` and `ArcInner` with `repr(C)`

Reply:
Could you also add a comment to those repr(C) annotations that notes that they're not for C-compat, but strictly so that the layout is equivalent for transmutable types (or so)?

Reply:
force-pushed